csrc/cpu/knn_cpu.cpp

```diff
@@ -0,0 +1,130 @@
+#include "radius_cpu.h"
+#include <algorithm>
+#include "utils.h"
+#include <cstdint>
+
+
+torch::Tensor knn_cpu(torch::Tensor support, torch::Tensor query, 
+			 int64_t k, int64_t n_threads){
+
+	CHECK_CPU(query);
+	CHECK_CPU(support);
+
+	torch::Tensor out;
+	std::vector<size_t>* neighbors_indices = new std::vector<size_t>(); 
+	auto options = torch::TensorOptions().dtype(torch::kLong).device(torch::kCPU);
+	int max_count = 0;
+
+	AT_DISPATCH_ALL_TYPES(query.scalar_type(), "radius_cpu", [&] {
+
+	auto data_q = query.data_ptr<scalar_t>();
+	auto data_s = support.data_ptr<scalar_t>();
+	std::vector<scalar_t> queries_stl = std::vector<scalar_t>(data_q,
+								   data_q + query.size(0)*query.size(1));
+	std::vector<scalar_t> supports_stl = std::vector<scalar_t>(data_s,
+								   data_s + support.size(0)*support.size(1));
+
+	int dim = torch::size(query, 1);
+
+	max_count = nanoflann_neighbors<scalar_t>(queries_stl, supports_stl ,neighbors_indices, 0, dim, 0, n_threads, k, 0);
+
+	});
+
+	size_t* neighbors_indices_ptr = neighbors_indices->data();
+
+	const long long tsize = static_cast<long long>(neighbors_indices->size()/2);
+	out = torch::from_blob(neighbors_indices_ptr, {tsize, 2}, options=options);
+	out = out.t();
+
+	auto result = torch::zeros_like(out);
+
+	auto index = torch::tensor({1,0});
+
+	result.index_copy_(0, index, out);
+
+	return result;
+}
+
+
+void get_size_batch(const std::vector<long>& batch, std::vector<long>& res){
+
+	res.resize(batch[batch.size()-1]-batch[0]+1, 0);
+	long ind = batch[0];
+	long incr = 1;
+	for(unsigned long i=1; i < batch.size(); i++){
+
+		if(batch[i] == ind)
+			incr++;
+		else{
+			res[ind-batch[0]] = incr;
+			incr =1;
+			ind = batch[i];
+		}
+	}
+	res[ind-batch[0]] = incr;
+}
+
+torch::Tensor batch_knn_cpu(torch::Tensor support,
+			       torch::Tensor query,
+			       torch::Tensor support_batch,
+			       torch::Tensor query_batch,
+			       int64_t k) {
```

Comment:
Why do we use `batch` instead of `ptr` here? Wouldn't that make `size` computation trivial?

Reply:
Now that you mentioned it, I do think this part of logic is redundant. because we already asserted that `query_batch_stl` is sorted.

Reply:
I take it back (I responded in a hurry and misunderstood what you were saying). batch and ptr are the same thing.